ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java

```diff
@@ -404,6 +510,201 @@ private EndpointResponse handlePullQuery(
     }
   }
 
+  // There's no good way to find out the current position of
+  // Streams's Consumers on specific partitions, so I made this
```

Comment:
I may be missing something, but could we use `ProcessorContext#partition` and `ProcessorContext#offset` and add a `process` step into the topology at the end? Another benefit of that is that we could have it stop processing any partitions that have passed their end-offset directly in there and not rely on something asynchronous to shut it down all-or-nothing

Reply:
Yeah, I did it that way first. I'm not sure if it wound up in any of the in-progress commits in this branch.

One problem is that a processor doesn't have the power to stop its own task or request shutdown of the runtime or anything like that. The only thing it could do is just start to drop all records it receives, but that alone doesn't help the query to terminate.

What did kind of work for me was passing a semaphore in to a processor at the end of the topology, which used the context to detect when it was "done" (just as you describe), and count-down the semaphore. Then, there was a thread that I hackily started in the StreamedQueryResource that blocked on the semaphore and then shut down the app. The problem I faced is that a processor at the end of the topology actually doesn't see all the records (for example in the test query I did with a `WHERE ` cause), so it won't know when Streams as a whole has passed the desired end point.

My conclusion was that there's not really a great way to implement the desired logic in the form of a processor, at least not without a significant rewrite of the way KSQL constructs the Streams topology.

Another alternative I considered was to change the "push query queue" so that the transient query would not just pipe out the GenericRow, but also include other metadata like the partition and offset, so we could decide in the QueryStreamWriter directly when we are "done". But that's also vulnerable to the fact that if we filter out some records, the writer isn't guaranteed to see all the offsets we've processed.

Reply:
Oh, boy. None of this was necessary. I was thinking that we couldn't use committed offsets, because "transient" queries wouldn't commit offsets... But of course transient queries are just regular Streams apps and must commit offsets, so we can use the committedOffsets API to know when Streams has passed the desired point, no magic required.